Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -3383,7 +3403,13 @@ private async Task<PipExecutionStep> ExecutePipStep(RunnablePip runnablePip)
                     }
 
                 case PipExecutionStep.MaterializeOutputs:
-                    {
+                    {
+                        if (m_configuration.Distribution.FireForgetMaterializeOutput && !AnyPendingPipsExceptMaterializeOutputs())
+                        {
+                            // There is no pips running anything except materializeOutputs. 
+                            m_schedulerCompletionExceptMaterializeOutputs.TrySetResult(false);

```

Comment:
>false [](start = 87, length = 5)

if one is looking only at this block of code, it's not clear / confusing why we are setting `m_schedulerCompletionExceptMaterializeOutputs` to false; i.e., essentially we have a bool that says 'completed but some materialization is still running', and once the engine enters this state, we set that bool to `!true`

Reply:
I understand that you setting it to false so you can put it through the existing logic (`Task.WhenAny(m_attachCompletion.Task, m_schedulerCompletion.Task)`) and I appreciate that you added that detailed comment there; but I think we need to make it clear what is going on here (in this method). Maybe we need to rename the variable?

Alternatively, we can set it here to 'true' so the var name aligns with the assigned value, and change the `RemoteWorker` :
```
var attachCompletionResult = await Task.WhenAny(m_attachCompletion.Task, m_schedulerCompletion.Task);

if (attachCompletionResult == m_schedulerCompletion.Task || !await attachCompletionResult) 
{
...
} 
``` 

---
In reply to: [348756422](https://github.com/microsoft/BuildXL/pull/1216#discussion_r348756422) [](ancestors = 348756422)